### ISSUE TITLE
How do I convert the raw data obtained by the microphone into the data required for the spectrogram?

### ISSUE THREAD

**Issue by tang158152** [liamappelbe/fftea#51]
<img width="384" alt="截屏2024-08-21 14 16 40" src="https://github.com/user-attachments/assets/a526ce75-510f-471b-a4ec-24aa0bc05fc7">

**Comment by tang158152** [comment]
How do I convert the raw data obtained by the microphone into the data required for the spectrogram?
I don't know much about audio, please forgive me

**Comment by liamappelbe** [comment]
Need some details about what the raw data looks like. What is the Dart type? How is it encoded?

Assuming it's just, say, an [`Int16List`](https://api.flutter.dev/flutter/dart-typed_data/Int16List-class.html), you just need to convert it to a `List<double>` or `Float64List` by mapping the int16 range (-32768 to 32767) to the usual range for float audio sample (-1 to 1). Ie, just divide each element by 32768.

Also, my experience with microphone libraries in flutter has been pretty bad. I'd recommend verifying the data you're getting from the microphone by saving it to a file and listening back to it, before you try to run it through a spectrogram. See [this comment](https://github.com/liamappelbe/fftea/issues/45#issuecomment-1848796230).

**Comment by tang158152** [comment]
> Need some details about what the raw data looks like. What is the Dart type? How is it encoded?
> 
> Assuming it's just, say, an [`Int16List`](https://api.flutter.dev/flutter/dart-typed_data/Int16List-class.html), you just need to convert it to a `List<double>` or `Float64List` by mapping the int16 range (-32768 to 32767) to the usual range for float audio sample (-1 to 1). Ie, just divide each element by 32768.
> 
> Also, my experience with microphone libraries in flutter has been pretty bad. I'd recommend verifying the data you're getting from the microphone by saving it to a file and listening back to it, before you try to run it through a spectrogram. See [this comment](https://github.com/liamappelbe/fftea/issues/45#issuecomment-1848796230).

Mine is a audio_streamer plug-in, I see that the data is sent every 0.5, 19200 data is delivered at a time, this is the data issued every 0.5 seconds, can you see the details of these raw data
[Uploading data.txt…]()

**Comment by tang158152** [comment]
And I pay attention to the big, in this 19,200 data, the previous data is always large

**Comment by tang158152** [comment]
[data.txt](https://github.com/user-attachments/files/16703664/data.txt)

**Comment by tang158152** [comment]
> [数据.txt](https://github.com/user-attachments/files/16703664/data.txt)

This is data acquired every 0.5 seconds

**Comment by liamappelbe** [comment]
yep, looks like it's already in the correct format. Have you tried running it through `STFT.stream`?

**Comment by tang158152** [comment]
> STFT.stream

Is there a suitable plugin library for STFT.stream, I don't know much about audio

**Comment by liamappelbe** [comment]
> Is there a suitable plugin library for STFT.stream, I don't know much about audio

Not sure what you mean. You can use this package anywhere you'd want to use a plugin. package:fftea is pure Dart, so it's fully compatible with all Flutter platforms. No plugin needed.
